### PR TITLE
refactor(moe): bind TRTLLM launch state to each call

### PR DESCRIPTION
**Pull request by feih-nv**
## 📌 Description

Previously, TRTLLM `pack_inputs()` returned a plain tensor list (`moe_inputs.to_list()`) while storing per-call launch metadata in shared `runner._static_kwargs` and the matching tuning configuration in `runner.tuning_config`. A later `pack_inputs()` call could overwrite both before an earlier call reached `forward()`.

This PR introduces `_TrtllmPackedInputs`, a list-compatible per-call bundle that carries the tensors, immutable `_TrtllmLaunchState`, and matching `TuningConfig` together. Consequently, `pack(A) → pack(B) → forward(A)` still executes with A’s launch state and tuning configuration.

Implementation details and validation:

- thread the paired launch state through `MoELayer`, the autotuner, and the unified MoE benchmark;
- keep token-dynamic `per_token_scale` in `MoeRunnerInputs`, where token-bucket synthesis can resize it, instead of double-sourcing it from static launch kwargs; and
- retain the architecture-independent synthetic `pack(A) -> pack(B) -> forward(A)` regression while validating the same packed-call contract in real supported-GPU packer tests, including rejection after conversion to a plain list.

The scope is intentionally TRTLLM-only. This does not make every `MoELayer` backend fully thread-safe: b12x retains shape-dependent wrapper/buffer ownership, while CUTLASS retains shared workspace allocation and tactic-selection state. Those follow-ups are tracked separately.

**Longer-term design:** A cleaner cross-backend abstraction would use composition: a `RunnerCall`/`Invocation` object with `inputs`, `launch_state`, and `tuning_config` as explicit members, consumed directly by both runners and the autotuner. This PR intentionally does not perform that broader migration because the existing `TunableRunner`, `AutoTuner`, and TRTLLM inner-runner interfaces consume and synthesize plain tensor lists. Changing that contract would touch every backend and tuning call site, substantially expanding the scope and regression risk. For example:

```
@dataclass(frozen=True)
class RunnerCall:
    inputs: list[Tensor]
    launch_state: LaunchState
    tuning_config: TuningConfig
```

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used my preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

Focused validation passed for:

- the architecture-independent synthetic interleaved-packed-call regression and TRTLLM fake-output contracts;
- real TRTLLM packer contract assertions on supported hardware;
- TRTLLM FP4 per-token CUDA-graph behavior;
- unified MoE autotuner dispatch with per-call launch metadata.

### CODE REVIEW

**Comment by feih-nv** [comment]
@coderabbitai review

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by feih-nv** [comment]
@flashinfer-bot run

File: tests/moe/test_unified_moe.py [hunk 1]
```diff
@@ -2952,12 +3063,15 @@ def _make_bf16_from_logits_inputs(self, logits_dtype):
 
     @pytest.mark.parametrize("logits_dtype", [torch.float32, torch.bfloat16])
     def test_bf16_expert_weights_buffer_is_bf16(self, logits_dtype):
-        runner, _, moe_inputs, logits = self._make_bf16_from_logits_inputs(logits_dtype)
+        runner, inputs, moe_inputs, logits = self._make_bf16_from_logits_inputs(
```
[the remark this answers is not in the document]
Comment (feih-nv) [reply tests/moe/test_unified_moe.py]: Fixed in dbbea64e4 by marking the intentionally unused runner binding as `_runner`.

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by feih-nv** [comment]
@flashinfer-bot run

**Comment by aleozlx** [comment]
## PR Review Screening

**CI verdict:** ✅ auto-run ok
**Review category:** `live` (rule fired: C3.2 durable — the `MoERunner` base class and `MoELayer`'s dispatch path adopt a new `pack_inputs` → `forward` calling convention)
**Blocking checks:** none
**Release blocker:** no
**Early stop:** no

### Security
| Q | Answer | Evidence |
|---|--------|----------|
| S1 injection/supply-chain | no | — |
| S2 template overwritten | no | default template structure retained |
| S3 template obligations | met | 5/5 checkboxes ticked and substantiated; Tests section lists the validation performed |
| S4 agent-directing text | no | — |

### Packaging
| Q | Answer | Evidence |
|---|--------|----------|
| C1.1 external dependency bump | no | no `supply`-bucket files touched |
| C1.2 public API changes | yes — **extension of existing shape** (with one breaking removal) | see signatures below |
| C1.3 AOT / trace registration | n-a | no `gen_*_module()` added, no new `@flashinfer_api` |

C1.2 — `flashinfer/fused_moe/runners.py` (public runner classes are exported in `flashinfer.fused_moe.__all__`):
- net-new on the shared base `MoERunner`: `tuning_config_for(inputs) -> TuningConfig`, `launch_state_for(inputs) -> Any`, `launch_kwargs_for(inputs) -> dict[str, Any]`
- `_TrtllmRunnerBase._forward_inner(inputs, tactic, do_preparation)` → `(..., launch_state: _TrtllmLaunchState | None = None)`; `Trtllm*Runner.forward()` now consumes a `launch_state` kwarg
- `Trtllm*Runner.pack_inputs()` return `List[torch.Tensor]` → `_TrtllmPackedInputs` (list subclass carrying `tuning_config` + `launch_state`)
- **removed**: `Trtllm*Runner.tuning_config` and `._static_kwargs` attributes — direct-runner callers must switch to `tuning_config_for(inputs)` / `launch_kwargs_for(inputs)`
- `flashinfer/fused_moe/layer.py`: `MoELayer` public signature unchanged

### Presentation
| Q | Answer | Evidence |
|---|--------|----------|
| C2.1 perf claim backed by data | n-a | refactor; no perf claim in title or description |

### Implementation
| Q | Answer | Evidence |
|---|--------|----------|
| C3.1 experimental track declared | no | labels `run-ci`, `op: moe`; no `flashinfer/experimental/` paths, no `@flashinfer_experimental_api` |
| C3.2 shared internals touched | yes — **durable** | `MoERunner` (base of the CUTLASS / CuteDSL / TRTLLM / b12x runners) plus `MoELayer.__call__` and `_select_winner` dispatch |
| C3.3 tests match behavior change | yes | arch-independent interleaved-pack regression + SM100 real-packer contract asserts via `tests/moe/utils.py::assert_trtllm_packed_call_contract` |
| C3.4 critical fix / release blocker | no | framed as a refactor; no linked issue, no hang/crash/corruption claim |

### Experimental track
| Q | Answer | Evidence |
|---|--------|----------|
| C4.1 isolated from common areas | n-a | C3.1 = no |

**Notes for the maintainer**
- Breaking for out-of-tree direct-runner callers: `runner.tuning_config` / `runner._static_kwargs` no longer exist on TRTLLM runners, and passing `list(inputs)` or a slice of a packed call now raises.
- The new convention is load-bearing for every backend: `MoELayer` calls `launch_kwargs_for()` on all runners, so any runner or test double that does not derive from `MoERunner` must implement it (see the CUTLASS test double edit in `tests/moe/test_unified_moe_cutlass.py`).
- Internal CI is not green yet — the most recent reported pipeline passed 15/16 test jobs.

<sub>Generated by flashinfer-pr-screen · rubric: docs/code_review_guidance.md · not a code review · AI screening can make mistakes — a maintainer's judgment supersedes this report.</sub>

**Review by aleozlx — APPROVED** [review]
lgtm
